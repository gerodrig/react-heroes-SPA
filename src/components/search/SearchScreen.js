import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import { useMemo } from 'react';

export const SearchScreen = () => {
	const navigate = useNavigate();
	const location = useLocation();

	const { q = '' } = queryString.parse(location.search);

	const [formValues, handleInputChange, reset] = useForm({
		searchText: q,
	});

	const { searchText } = formValues;

	const heroesFiltered = useMemo( () => getHeroesByName(q), [q]);

	const handleSearch = (e) => {
		e.preventDefault();
		//print search Text
		navigate(`?q=${searchText}`);
		reset();
	};

	return (
		<>
			<h1>Hero Finder</h1>
			<hr />

			<div className='row'>
				<div className='col-5 animate__animated animate__fadeIn animate__slow'>
					<h4>Search</h4>
					<hr />

					<form onSubmit={handleSearch}>
						<input
							type='text'
							name='searchText'
							className='form-control'
							autoComplete='off'
							placeholder='Search for a hero'
							value={searchText}
							onChange={handleInputChange}
							required
						/>

						<button
							className='btn btn-primary btn-block mt-2'
							type='submit'
						>
							Search
						</button>
					</form>
				</div>
				<div className='col-7 animate__animated animate__fadeIn animate__slow'>
					<h4>Results</h4>
					<hr />

					{q === '' ? (
						<div className='alert alert-info animate__animated animate__bounceInDown'>
							Search for a Hero
						</div>
					) : (
						heroesFiltered.length === 0 && (
							<div className='alert alert-danger animate__animated animate__bounceInDown'>
								No results found for { q }
							</div>
						)
					)}

					{heroesFiltered.map((hero) => (
						<HeroCard key={hero.id} {...hero} />
					))}
				</div>
			</div>
		</>
	);
};
