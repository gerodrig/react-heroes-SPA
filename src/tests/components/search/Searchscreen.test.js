import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockNavigate,
}));

describe('Tests in <SearchScreen /> component', () => {
	test('Should render search snapshot correctly with default values', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('.alert-info').text().trim()).toBe(
			'Search for a Hero'
		);
	});

	test('should render Batman and the input should contain the queryString', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(wrapper.find('input').prop('value')).toBe('batman');
	});

	test('should show an error if the hero is not found', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search?q=batman123']}>
				<SearchScreen />
			</MemoryRouter>
		);

		expect(wrapper.find('.alert-danger').text().trim()).toBe(
			'No results found for batman123'
		);
	});

	test('should call the navigate to the new pathname', () => {
		const wrapper = mount(
			<MemoryRouter initialEntries={['/search']}>
				<SearchScreen />
			</MemoryRouter>
		);

        wrapper
            .find('input')
            .simulate('change', {
                target: { name: 'searchText', value: 'batman' },
            });
        act(() => {
    
            wrapper.find('form').prop('onSubmit')({ preventDefault(){} });
        });

		expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
	});
});
