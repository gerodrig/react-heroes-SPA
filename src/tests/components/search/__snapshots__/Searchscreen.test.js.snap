// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen /> component Should render search snapshot correctly with default values 1`] = `
Array [
  <h1>
    Hero Finder
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5 animate__animated animate__fadeIn animate__slow"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search for a hero"
          required={true}
          type="text"
          value=""
        />
        <button
          className="btn btn-primary btn-block mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 animate__animated animate__fadeIn animate__slow"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__bounceInDown"
      >
        Search for a Hero
      </div>
    </div>
  </div>,
]
`;
